x/ccv/provider/keeper/punish_validators_test.go

```diff
@@ -0,0 +1,136 @@
+			) []*gomock.Call {
+				return []*gomock.Call{
+					// We only expect a single call to GetValidatorByConsAddr.
+					mocks.MockStakingKeeper.EXPECT().GetValidatorByConsAddr(
+						ctx, providerConsAddr.ToSdkConsAddr()).Return(
+						stakingtypes.Validator{Status: stakingtypes.Unbonded}, true,
+					).Times(1),
+				}
+			},
+		},
+		{
+			"tombstoned validator",
+			providerConsAddr,
+			func(ctx sdk.Context, mocks testkeeper.MockedKeepers,
+				provAddr types.ProviderConsAddress,
+			) []*gomock.Call {
+				return []*gomock.Call{
+					mocks.MockStakingKeeper.EXPECT().GetValidatorByConsAddr(
+						ctx, providerConsAddr.ToSdkConsAddr()).Return(
+						stakingtypes.Validator{}, true,
+					).Times(1),
+					mocks.MockSlashingKeeper.EXPECT().IsTombstoned(
+						ctx, providerConsAddr.ToSdkConsAddr()).Return(
+						true,
+					).Times(1),
+				}
+			},
+		},
+		{
+			"jailed validator",
+			providerConsAddr,
+			func(ctx sdk.Context, mocks testkeeper.MockedKeepers,
+				provAddr types.ProviderConsAddress,
+			) []*gomock.Call {
+				return []*gomock.Call{
+					mocks.MockStakingKeeper.EXPECT().GetValidatorByConsAddr(
+						ctx, providerConsAddr.ToSdkConsAddr()).Return(
+						stakingtypes.Validator{Jailed: true}, true,
+					).Times(1),
+					mocks.MockSlashingKeeper.EXPECT().IsTombstoned(
+						ctx, providerConsAddr.ToSdkConsAddr()).Return(
+						false,
+					).Times(1),
+					mocks.MockSlashingKeeper.EXPECT().JailUntil(
+						ctx, providerConsAddr.ToSdkConsAddr(), evidencetypes.DoubleSignJailEndTime).
+						Times(1),
+					mocks.MockSlashingKeeper.EXPECT().Tombstone(
+						ctx, providerConsAddr.ToSdkConsAddr()).
+						Times(1),
+				}
+			},
+		},
+		{
+			"bonded validator",
+			providerConsAddr,
+			func(ctx sdk.Context, mocks testkeeper.MockedKeepers,
+				provAddr types.ProviderConsAddress,
+			) []*gomock.Call {
+				return []*gomock.Call{
+					mocks.MockStakingKeeper.EXPECT().GetValidatorByConsAddr(
+						ctx, providerConsAddr.ToSdkConsAddr()).Return(
+						stakingtypes.Validator{Status: stakingtypes.Bonded}, true,
+					).Times(1),
+					mocks.MockSlashingKeeper.EXPECT().IsTombstoned(
+						ctx, providerConsAddr.ToSdkConsAddr()).Return(
+						false,
+					).Times(1),
+					mocks.MockStakingKeeper.EXPECT().Jail(
+						ctx, providerConsAddr.ToSdkConsAddr()).
+						Times(1),
+					mocks.MockSlashingKeeper.EXPECT().JailUntil(
+						ctx, providerConsAddr.ToSdkConsAddr(), evidencetypes.DoubleSignJailEndTime).
+						Times(1),
+					mocks.MockSlashingKeeper.EXPECT().Tombstone(
+						ctx, providerConsAddr.ToSdkConsAddr()).
+						Times(1),
+				}
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		providerKeeper, ctx, ctrl, mocks := testkeeper.GetProviderKeeperAndCtx(
+			t, testkeeper.NewInMemKeeperParams(t))
+
+		// Setup expected mock calls
+		gomock.InOrder(tc.expectedCalls(ctx, mocks, tc.provAddr)...)
+
+		// Execute method and assert expected mock calls
+		providerKeeper.JailAndTombstoneValidator(ctx, tc.provAddr)
```

Comment:
if a test fails would we be able to easily see the `tc.name` for the failed test?

Reply:
Not sure! Maybe it's worth finding a way to print the subtests' name like in the integration tests:
`func (*suite.Suite).Run(name string, subtest func()) bool`.